relayer/chains/evm/listener.go

```diff
@@ -54,25 +51,17 @@ func (p *Provider) Listener(ctx context.Context, lastSavedHeight uint64, blockIn
 
 	p.log.Info("Start from height ", zap.Uint64("height", startHeight), zap.Uint64("finality block", p.FinalityBlock(ctx)))
 
-	subscribeStart := time.NewTicker(time.Second * 1)
-	pollHeightTicker := time.NewTicker(time.Second * 1)
-
-	pollHeightTicker.Stop()
-
-	resetFunc := func() {
-		subscribeStart.Reset(time.Second * 3)
-		pollHeightTicker.Reset(time.Second * 3)
-	}
+	var (
+		subscribeStart = time.NewTicker(time.Second * 1)
+		isSubError     bool
```

Comment:
once this value becomes true after subscription error, I think it should be made false once subscription is working normally. otherwise, there is possibility of catching same events from subscription and polling and relayer will try to deliver same message twice.

Reply:
nope. `subError` truthiness is used after relayer is started and it is also used to determine if it's not the first time relayer is started, otherwise `subError` will be always true. making it `false` does not do anything.

Reply:
Ohh I see, subscribeStart ticker runs only once. 
```go
go p.Subscribe(ctx, blockInfoChan, resetFunc)

if isSubError {
startHeight = p.GetLastSavedBlockHeight()
}
```
SO, in this above block since Subscribe is done asynchronously, the next check of `if isSubError` actually runs before Subscribe function. So, this is not what we want, right? 

Also next thing, the websocket error can occur at any time in the runtime. That part is not handled. If websocket connection gets disconnected, I don't see any strategy to reconnect and poll for the missing block due to disconnection.

Also there is deadlock in this code. I wrote similar code and tried to run, I got deadlock.
```go
func main() {
	ticker := time.NewTicker(time.Second * 1)
	for {
		select {
		case <-ticker.C:
			ticker.Stop()
			var wg sync.WaitGroup

			fmt.Println("ticking each second")

			for i := 0; i < 5; i++ {
				wg.Add(1)
				fmt.Println("andar")
				go func(wg *sync.WaitGroup) {
					defer wg.Done()
					time.Sleep(5 * time.Second)
				}(&wg)
			}
			go func() {
				wg.Wait()
			}()
		}
	}
}
```

Reply:
sorry to bother about deadlock issue, in the snippet I provided, the deadlock was because of a single case. But another concern is since the ticker runs only once, the for loop is taking CPU usage for no work infinitely; better to return or break the for loop once the job inside the ticker is completed.

Reply:
We do have these all the problems you mentioned already.